code/bngblaster/src/bbl_protocols.c

```diff
@@ -2202,12 +2202,40 @@ decode_icmpv6(uint8_t *buf, uint16_t len,
             }
             break;
         case IPV6_ICMPV6_NEIGHBOR_SOLICITATION:
+            if(len < 20) {
+                return DECODE_ERROR;
+            }
+            BUMP_BUFFER(buf, len, sizeof(uint32_t)); /* flags / reserved */
+            memcpy(&icmpv6->prefix.address, buf, IPV6_ADDR_LEN);
+            break;
         case IPV6_ICMPV6_NEIGHBOR_ADVERTISEMENT:
             if(len < 20) {
                 return DECODE_ERROR;
             }
             BUMP_BUFFER(buf, len, sizeof(uint32_t)); /* flags / reserved */
             memcpy(&icmpv6->prefix.address, buf, IPV6_ADDR_LEN);
+            BUMP_BUFFER(buf, len, IPV6_ADDR_LEN);
+            while(len >= 2) {
+                option = *buf;
+                BUMP_BUFFER(buf, len, sizeof(uint8_t));
+                option_len = (*buf) * 8;
+                BUMP_BUFFER(buf, len, sizeof(uint8_t));
+                if(option_len < 2 || len + 2 < option_len) {
+                    return DECODE_ERROR;
+                }
+                if(option == ICMPV6_OPTION_DEST_LINK_LAYER) {
+                    if(option_len != 8) {
+                        // Maleformed ICMPv6 packet
+                        return DECODE_ERROR;
+                    }
+                    icmpv6->dst_mac = buf;
+                    break;
+                }
+                if(option_len > len) {
+                    return DECODE_ERROR;
+                }
```

Comment:
this check is not needed because we already check that `len < option_len-2`

Reply:
So `option_len` can still be bigger than `len` e.g. `len = 7; option_len = 8`.

With your request from above this is partially handled by `while(len >= 2)`, but removing this snipped would accept icmpv6 packets where there is a trailing byte.

FWIW because of the `while(len >= 2)`, we would still accept icmpv6 packets with a trailing byte when there are no options.

I don't mind being less strict for the bngblaster use cases, but please confirm if this is wanted.

Reply:
removed, also I've changed the while condition to `len >= 8`, icmpv6 options must be a multiple of 8, so that's the smallest valid option